### PR TITLE
chore(core): unify address type for socket-based component configs

### PR DESCRIPTION
**Pull request by tobz**
This PR seeks to start the larger process of unifying codepaths that are shared between related component types. The initial cut of this is reworking some of the socket-based components (`socket` source/sink, `statsd` source, `syslog` source, etc) to all use `SocketListenAddr` for describing their listen address, as well as unifying the internal events they emit.

Practically speaking, our introduction of `SocketListenAddr` -- which makes it possible to specify a classic listen address such as `localhost:5000` _or_ the offset of a socket managed by systemd's socket activation feature -- focused on usage for components looking to listen on a TCP port. However, systemd's socket activation feature has support for UDP sockets. More tactically, however, is the fact that the status quo generates a configuration schema for these components which is hard to present a unified output for, as between TCP and UDP, it appears that the "address" fields have different schemas, even though one of them (TCP) is merely a superset.

By using `SocketListenAddr` for both TCP and UDP, we can both expose existing support to make these components more useful, as well as fix the issue at hand with their resulting configuration schema.

Additionally, I've unified some of the basic internal events: failure to bind, failure to receive, and failure to send. The underlying issue is the same across all of these components, but we had varying error messages, error codes, and error types, for no good reason.

### CODE REVIEW

File: src/internal_events/udp.rs [hunk 1]
```diff
@@ -4,6 +4,8 @@ use vector_core::internal_event::InternalEvent;
 
 use crate::{emit, internal_events::SocketOutgoingConnectionError};
 
+// TODO: Get rid of this. UDP is connectionless, so there's no "successful" connect event, only
+// successfully binding a socket that can be used for receiving.
```
Comment (spencergilbert) [inline src/internal_events/udp.rs:8]: What's blocking these TODOs today?

**Review by spencergilbert — APPROVED** [review]
👍 like the change, couple CI problems for unused funcs though.

File: src/internal_events/udp.rs (hunk 1 above)
Comment (tobz) [reply src/internal_events/udp.rs:8]: Nothing besides me trying to keep this PR smaller.

Comment (spencergilbert) [reply src/internal_events/udp.rs:8]: That was my assumption, but wanted to double check!

**Pull request merged**